source/common/http/filter/ext_authz.cc

```diff
@@ -0,0 +1,126 @@
+#include "common/http/filter/ext_authz.h"
+
+#include <string>
+#include <vector>
+
+#include "envoy/http/codes.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/ext_authz/ext_authz_impl.h"
+#include "common/http/codes.h"
+#include "common/router/config_impl.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace Http {
+namespace ExtAuthz {
+
+namespace {
+
+static const Http::HeaderMap* getDeniedHeader() {
+  static const Http::HeaderMap* header_map = new Http::HeaderMapImpl{
+      {Http::Headers::get().Status, std::to_string(enumToInt(Code::Forbidden))}};
+  return header_map;
+}
+
+} // namespace
+
+void Filter::initiateCall(const HeaderMap& headers) {
+
+  Router::RouteConstSharedPtr route = callbacks_->route();
+  if (route == nullptr || route->routeEntry() == nullptr) {
+    return;
+  }
+
+  const Router::RouteEntry* route_entry = route->routeEntry();
+  Upstream::ThreadLocalCluster* cluster = config_->cm().get(route_entry->clusterName());
+  if (cluster == nullptr) {
+    return;
+  }
+  cluster_ = cluster->info();
+
+  Envoy::ExtAuthz::CheckRequestUtils::createHttpCheck(callbacks_, headers, checkRequest_);
+
+  state_ = State::Calling;
+  initiating_call_ = true;
+  client_->check(*this, checkRequest_, callbacks_->activeSpan());
+  initiating_call_ = false;
+}
+
+FilterHeadersStatus Filter::decodeHeaders(HeaderMap& headers, bool) {
+  initiateCall(headers);
+  return (state_ == State::Calling) ? FilterHeadersStatus::StopIteration
+                                    : FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus Filter::decodeData(Buffer::Instance&, bool) {
+  return (state_ == State::Calling) ? FilterDataStatus::StopIterationAndWatermark
+                                    : FilterDataStatus::Continue;
+}
+
+FilterTrailersStatus Filter::decodeTrailers(HeaderMap&) {
+  return state_ == State::Calling ? FilterTrailersStatus::StopIteration
+                                  : FilterTrailersStatus::Continue;
+}
+
+void Filter::setDecoderFilterCallbacks(StreamDecoderFilterCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+}
+
+void Filter::onDestroy() {
+  if (state_ == State::Calling) {
+    state_ = State::Complete;
+    client_->cancel();
+  }
+}
+
+void Filter::onComplete(Envoy::ExtAuthz::CheckStatus status) {
+  ASSERT(cluster_);
+
+  state_ = State::Complete;
+
+  using Envoy::ExtAuthz::CheckStatus;
+
+  switch (status) {
+  case CheckStatus::OK:
+    cluster_->statsScope().counter("ext_authz.ok").inc();
```

Comment:
I'm wondering if we want to charge to cluster or route? Why charge to cluster here but globally in the TCP ext_authz?

Reply:
@htuch To be honest I am not sure i fully understand the implications/scope of this choice. I was following the stats model from the `ratelimit` filter (but u know that already).
From what I understood of the flow i thought that since the cluster was the routed destination hence we were accounting towards it.